File: _misc/2022/tarai/ruby/tarai.rb
require 'benchmark'

def tarai(x, y, z)
  if x > y
    tarai(tarai(x-1, y, z), tarai(y-1, z, x), tarai(z-1, x, y))
  else
    y
  end
end

arr = [[10, 5], [11, 5], [12, 6], [13, 6], [13, 7], [14, 7]]

arr.each do |x, y|
  puts "x = #{x}, y = #{y}"
  result = (1..10).map do
    Benchmark.measure { tarai(x, y, 0) }
  end
  puts Benchmark::CAPTION
  puts result.sum(Benchmark::Tms.new) / result.size
end

File: _misc/2022/tarai/python/tarai.py
import timeit
import pandas as pd

def tarai(x, y, z):
    if x > y:
        return tarai(tarai(x-1, y, z), tarai(y-1, z, x), tarai(z-1, x, y))
    else:
        return y

arr = [(10, 5), (11, 5), (12, 6), (13, 6), (13, 7), (14, 7)]
for x, y in arr:
    print("\nx = %d, y = %d" % (x, y))
    result = timeit.repeat(lambda: tarai(x, y, 0), number=1, repeat=10)
    print(result)
    print(pd.DataFrame(result).describe().transpose())
